src/lib.rs

```diff
@@ -556,10 +663,117 @@ impl<T> Receiver<T> {
+                                _ => unreachable!(),
+                            }
+                        }
+                        None => {
+                            // ORDERING: synchronize with the write of the message
+                            match channel.state.swap(EMPTY, Acquire) {
+                                // We reached the end of the timeout without receiving a message
+                                RECEIVING => {
+                                    // SAFETY: we were in th receiving state and are now in the empty
+                                    // state, so the sender has not and will not try to read the waker,
+                                    // so we have exclusive access to drop it.
+                                    unsafe { channel.drop_waker() };
+
+                                    break Err(RecvTimeoutError::Timeout);
+                                }
+                                // The sender sent the message while we were parked.
+                                MESSAGE => {
+                                    // Same safety and ordering as the Some branch
+
+                                    channel.state.store(DISCONNECTED, Relaxed);
+                                    break Ok(unsafe { channel.take_message() });
+                                }
+                                // The sender was dropped while we were parked.
+                                DISCONNECTED => {
+                                    // ORDERING: we were originally in the disconnected state meaning
+                                    // that the sender is inactive and no longer observing the state,
+                                    // so we only need to change it back to DISCONNECTED for if the
+                                    // receiver is dropped or a recv* method is called again
+                                    channel.state.store(DISCONNECTED, Relaxed);
+
+                                    break Err(RecvTimeoutError::Disconnected);
+                                }
+                                UNPARKING => {
+                                    // We were in the UNPARKING state and are now in the EMPTY state.
+                                    // We wait to be unparked since this is the only way to maintain
+                                    // correctness with intrusive memory.
 
-        match channel.state.load(SeqCst) {
+                                    break wait_for_unpark(channel);
+                                }
+                                _ => unreachable!(),
+                            }
+                        }
+                    }
+                }
+            },
+        )
+    }
+
+    #[cfg(feature = "std")]
+    #[inline(always)]
+    fn start_recv_ref<E>(
+        &self,
+        disconnected: impl FnOnce() -> E,
+        finish: impl FnOnce(&Channel<T>) -> Result<T, E>,
+    ) -> Result<T, E> {
```

Comment:
I think some documentation on roughly what `start_recv_ref` is doing would be great for understanding this code.